pkg/server/image_pull.go

```diff
@@ -127,7 +128,34 @@ func (c *criService) PullImage(ctx context.Context, r *runtime.PullImageRequest)
 			containerd.WithChildLabelMap(containerdimages.ChildGCLabelsFilterLayers))
 	}
 
-	image, err := c.client.Pull(ctx, ref, pullOpts...)
+	// Check if there's currently an ongoing pull of the same image, and if so just wait on the result.
+	if c.config.DisableSameImageParallelPull {
+		bc, ok := c.imagesInProgress.load(ref)
```

Comment:
I think adding a `LoadOrStore` method on `inProgress` and using that would be clearer. That way this can just be written as: 
```
bc, ok := c.imagesInProgress.loadOrStore(ref, newImageBoradCaster())
if ok {
   bc.wait()
   ....
}
```

Reply:
I think I had tried or thought of this but I stopped for some reason.. let me try and trace back my thoughts. Agreed that does sound like a more sound approach

Reply:
Changed to this, thanks for the suggestion!